test/suite/lib.rs

```diff
@@ -218,13 +219,84 @@ pub async fn test_supported_standards(ledger: impl LedgerEnv) -> anyhow::Result<
     Ok(Outcome::Passed)
 }
 
+/// Checks whether the ledger applies deduplication of transactions correctly
+pub async fn test_tx_deduplication(ledger_env: impl LedgerEnv) -> anyhow::Result<Outcome> {
+    // Create two test accounts and transfer some tokens to the first account
+    let p1_env = setup_test_account(&ledger_env, 100_000.into()).await?;
+    let p2_env = p1_env.fork();
+
+    // If created at time is not set, the transfer should not be calssified as a duplicate --> Transfer should succeed
+    let transfer_args = Transfer::amount_to(10_000, p2_env.principal());
+    if transfer(&p1_env, transfer_args.clone())
+        .await
+        .unwrap()
+        .unwrap()
+        > transfer(&p1_env, transfer_args.clone())
+            .await
+            .unwrap()
+            .unwrap()
+    {
+        return Err(anyhow::Error::msg(
+            "BlockIndex of previous transaction is higher than that of the following transaction",
+        ));
+    }
+
+    let now = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap()
+        .as_nanos();
+    // Changing the timestamp to the previous transfer args should result in no deduplication --> Transfer should succeed
+    let transfer_args = transfer_args.created_at_time(now as u64);
+    transfer(&p1_env, transfer_args.clone())
+        .await
+        .unwrap()
+        .unwrap();
+    // Sending the previous transaction with the same timestamp twice should not be possible --> Transfer should not succeed
+    assert!(transfer(&p1_env, transfer_args.clone())
+        .await
+        .unwrap()
+        .is_err());
+
+    // Changing the fee to the previous transfer args should result in no deduplication --> Transfer should succeed
+    let transfer_args = transfer_args.fee(transfer_fee(&ledger_env).await.unwrap());
+    transfer(&p1_env, transfer_args.clone())
+        .await
+        .unwrap()
+        .unwrap();
+
+    // If we send the transfer again it should be a duplicate --> Transfer should not succeed
+    assert!(transfer(&p1_env, transfer_args.clone())
+        .await
+        .unwrap()
+        .is_err());
+
+    // Changing the memo to the previous transfer args should result in no deduplication --> Transfer should succeed
+    let transfer_args = transfer_args.memo(vec![1, 2, 3]);
+    transfer(&p1_env, transfer_args.clone())
+        .await
+        .unwrap()
+        .unwrap();
+
+    // If we send the transfer again it should be a duplicate --> Transfer should not succeed
+    assert!(transfer(&p1_env, transfer_args.clone())
+        .await
+        .unwrap()
+        .is_err());
+
```

Comment:
We could also test whether the ledger correctly identifies that transfer to (principal, None) and (principal, Some([0; 32])) are different transactions and are not deduplicated.

Reply:
Good idea